content/blog/deploying-mysql-schemas-using-dynamic-providers/index.md

```diff
@@ -0,0 +1,284 @@
+
+```python
+class SchemaInputs(object):
+    creator_name: Input[str]
+    creator_password: Input[str]
+    server_address: Input[str]
+    database_name: Input[str]
+    creation_script: Input[str]
+    deletion_script: Input[str]
+    def __init__(self, creator_name, creator_password, server_address, database_name, creation_script, deletion_script):
+        self.creator_name = creator_name
+        self.creator_password = creator_password
+        self.server_address = server_address
+        self.database_name = database_name
+        self.creation_script = creation_script
+        self.deletion_script = deletion_script
+```
+
+The second step is to write the Dynamic Provider itself. The provider handles all the create, read, update, and delete operations the resource needs. The `create` function creates a new resource and assigns it a unique ID, the `delete` function deletes an existing resource, the `diff` function determines if we can update the resource without having to fully replace it, and the `update` function performs those updates.
+
+```python
+class SchemaProvider(ResourceProvider):
+    def create(self, args):
+        connection = connector.connect(user=args["creator_name"],
+        password=args["creator_password"],
+        host=args["server_address"],
+        database=args["database_name"])
+        cursor = connection.cursor()
+        cursor.execute(args["creation_script"])
+        return CreateResult("schema-"+binascii.b2a_hex(os.urandom(16)).decode("utf-8"), outs=args)
+
+    def delete(self, id, args):
+        connection = connector.connect(user=args["creator_name"],
+        password=args["creator_password"],
+        host=args["server_address"],
+        database=args["database_name"])
+        cursor = connection.cursor()
+        cursor.execute(args["deletion_script"])
+
+    def diff(self, id, oldInputs, newInputs):
+        replaces = []
+        if (oldInputs["server_address"] != newInputs["server_address"]): replaces.append("server_address")
+        if (oldInputs["database_name"] != newInputs["database_name"]): replaces.append("database_name")
+        if (oldInputs["creation_script"] != newInputs["creation_script"]): replaces.append("creation_script")
+
+        return DiffResult(
+            changes=oldInputs != newInputs,
+            replaces=replaces,
+            stables=None,
+            delete_before_replace=True)
+
+    def update(self, id, oldInputs, newInputs):
+        return UpdateResult(outs={**newInputs})
+```
+
+And lastly, is the main Schema resource that we instantiate in our infrastructure code. We make all the inputs to our provider availible as outputs, so that they can be accessed and exported.
```

Comment:
```suggestion
And lastly, we add the main Schema resource that we instantiate in our infrastructure code. We make all the inputs to our provider available as outputs, so that they can be accessed and exported.
```